### PR TITLE
Change attachment InputStream to ChannelProvider

### LINKED ISSUE #674: Add a streamed multipart request
At the moment, Kord loads the entirety of an InputStream onto memory by [this `readBytes` call](https://github.com/kordlib/kord/blob/e3d97c249cd0c058c22a1b0a6cd1ddc729a3ea7c/rest/src/main/kotlin/request/Request.kt#L90). Since the `Request` class is sealed, one cannot make a custom request that converts just takes a `ByteReadChannel` (or rather a [`ChannelProvider`](https://github.com/ktorio/ktor/blob/main/ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/formDsl.kt#L135)) and streams file contents asynchronously.    
In fact appending a `ByteArray` form data part just adds another indirection, as at a lower-level, Ktor only has two types of parts: [Input or Channel](https://github.com/ktorio/ktor/blob/413bf1e30aaf063bef023e89fbf1c7b572b0163e/ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt#L144-L151) and uses `copyTo` on both, which asynchronously reads from one channel to another.   
Since making `ByteReadChannel`s or `Input`s is trivial, it would be nice change.

### PR DESCRIPTION
**Pull request by Tmpod**
Also remove full eager consuming of stream/channel when building request and
just pass it directly.

Closes #674 (discussed in [this](https://discord.com/channels/556525343595298817/1014497904116904067) Discord topic as well)

### CODE REVIEW

**Comment by Tmpod** [comment]
right, this is a binary incompatible change

**Comment by Lukellmann** [comment]
You need to run `./gradlew apiDump` and commit the changes to make the tests pass

**Review by Lukellmann — CHANGES_REQUESTED** [review]

File: rest/src/main/kotlin/NamedFile.kt [hunk 1]
```diff
@@ -1,12 +1,12 @@
 package dev.kord.rest
 
-import java.io.InputStream
+import io.ktor.utils.io.*
 
-public class NamedFile(public val name: String, public val inputStream: InputStream) {
+public class NamedFile(public val name: String, public val readChannel: ByteReadChannel) {
```
[the remark this answers is not in the document]
Comment (Tmpod) [reply rest/src/main/kotlin/NamedFile.kt:5]: oh yeah, my bad!

File: rest/src/main/kotlin/builder/message/create/MessageCreateBuilder.kt [hunk 2]
```diff
@@ -62,9 +61,8 @@ public sealed interface MessageCreateBuilder {
     /**
      * Adds a file with the given [path] to the attachments.
      */
-    public suspend fun addFile(path: Path): NamedFile = withContext(Dispatchers.IO) {
-        addFile(path.fileName.toString(), Files.newInputStream(path))
-    }
+    public suspend fun addFile(path: Path): NamedFile =
+        addFile(path.fileName.toString(), File(path.toUri()).readChannel())
```
[the remark this answers is not in the document]
Comment (Tmpod) [reply rest/src/main/kotlin/builder/message/create/MessageCreateBuilder.kt:65]: oops missed that, will fix

File: rest/src/main/kotlin/NamedFile.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tmpod) [reply rest/src/main/kotlin/NamedFile.kt:5]: I'm unsure what to do with `NamedFile`. Just provide a secondary constructor that accepts an `InputStream`? Then it wouldn't be fully binary compatible, would it? Since `inputStream` still wouldn't be a member and `component2` also does not return the same thing as before.  
Sorry, I'm not well versed in maintaining binary compat in Kotlin libraries.

[the remark this answers is not in the document]
Comment (Lukellmann) [reply rest/src/main/kotlin/NamedFile.kt:5]: you'll need:
- secondary constructor
- `inputStream` property with getter converting the `ByteReadChannel` to `InpuStream`
- hidden component2:
```kotlin
@Deprecated("Binary compatibility", level = HIDDEN)
@JvmName("component2")
public fun _component2(): InputStream = inputStream
```

**Comment by Tmpod** [comment]
Ooh, that's what HIDDEN is for, thanks!

File: rest/src/main/kotlin/NamedFile.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tmpod) [reply rest/src/main/kotlin/NamedFile.kt:5]: ooh, that's what HIDDEN is for, thanks!

File: rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt [hunk 3]
```diff
@@ -32,8 +32,19 @@ public sealed interface MessageModifyBuilder {
...
     public var attachments: MutableList<DiscordAttachment>?
 
+    /**
+     * Adds a file with the [name] and [content] to the attachments.
+     */
+    @Deprecated(
+        "Use ByteReadChannel instead of InputStream",
+        ReplaceWith(
+            "addFile(name, content.toByteReadChannel())",
+            "io.ktor.util.cio.toByteReadChannel"
```
Comment (Lukellmann) [inline rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt:42]:
```suggestion
            "io.ktor.utils.io.jvm.javaio.toByteReadChannel"
```
see above

File: rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt [hunk 4]
```diff
@@ -6,10 +6,10 @@ import dev.kord.rest.builder.component.ActionRowBuilder
 import dev.kord.rest.builder.component.MessageComponentBuilder
 import dev.kord.rest.builder.message.AllowedMentionsBuilder
 import dev.kord.rest.builder.message.EmbedBuilder
-import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
+import io.ktor.util.cio.*
+import io.ktor.utils.io.*
+import java.io.File
```
Comment (Lukellmann) [inline rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt:11]:
```suggestion
import io.ktor.util.cio.*
import io.ktor.utils.io.*
import io.ktor.utils.io.jvm.javaio.toByteReadChannel
import java.io.File
```
see above

File: rest/src/main/kotlin/request/Request.kt [hunk 5]
```diff
@@ -89,19 +90,13 @@ public class MultipartRequest<B : Any, R>(
...
-            }
-        } finally {
-            files.forEach { it.inputStream.close() }
+        files.forEachIndexed { index, pair ->
+            val (name, channel) = pair
+            append(
+                "file$index",
+                ChannelProvider { channel },
+                Headers.build { append(HttpHeaders.ContentDisposition, "filename=$name") }
+            )
```
Comment (Lukellmann) [inline rest/src/main/kotlin/request/Request.kt:99]:
```suggestion
        files.forEachIndexed { index, (fileName, fileChannel) ->
            append(
                "file$index",
                ChannelProvider { fileChannel },
                headersOf(HttpHeaders.ContentDisposition, "filename=$fileName"),
            )
```
while we are here we can clean this up a bit

File: rest/src/test/kotlin/request/MessageRequests.kt [hunk 6]
```diff
@@ -1,42 +0,0 @@
...
-import kotlinx.coroutines.runBlocking
-import org.junit.jupiter.api.MethodOrderer
-import org.junit.jupiter.api.Test
-import org.junit.jupiter.api.TestInstance
-import org.junit.jupiter.api.TestMethodOrder
-import java.io.InputStream
-
-@TestMethodOrder(MethodOrderer.MethodName::class)
-@TestInstance(TestInstance.Lifecycle.PER_CLASS)
-class MessageRequests {
```
Comment (Lukellmann) [inline rest/src/test/kotlin/request/MessageRequests.kt:16]: Could (and should) we bring this test back but with `assert(!readChannel.isClosedForWrite)` before and `assert(readChannel.isClosedForRead)` after executing a mock request (since the whole point is that the channel is consumed lazily)?

**Review by Lukellmann — CHANGES_REQUESTED** [review]

**Comment by Lukellmann** [comment]
Another thought I had was to use `ChannelProvider` directly as inputs for the functions (instead of `ByteReadChannel`). This would mean the channel wouldn't even have to be opened until executing the request. What is your thought on this?

(and btw sorry for requesting so many changes)

**Comment by Tmpod** [comment]
> Another thought I had was to use `ChannelProvider` directly as inputs for
> the functions (instead of `ByteReadChannel`). This would mean the channel
> wouldn't even have to be opened until executing the request. What is your
> thought on this?

Yeah I thought about it when first implementing this change, but ended up
deciding that taking a read channel directly would be more convenient most of
the time. However, also accepting a channel provider would make this even
more versatile (which was kinda the whole point to this change). Perhaps I
could change the internals once again to `ChannelProvider` and write
overloads for those `addFile` methods and such that can accept both providers
and channels directly (which would be wrapped in simple providers)?

> (and btw sorry for requesting so many changes)

No problem at all! In fact, I appreciate the detailed reviewing :)

**Comment by Lukellmann** [comment]
> However, also accepting a channel provider would make this even more versatile (which was kinda the whole point to this change). Perhaps I could change the internals once again to `ChannelProvider` and write overloads for those `addFile` methods and such that can accept both providers and channels directly (which would be wrapped in simple providers)?

Sounds good 👍

File: rest/src/main/kotlin/builder/message/create/MessageCreateBuilder.kt [hunk 7]
```diff
@@ -53,7 +55,24 @@ public sealed interface MessageCreateBuilder {
     /**
      * Adds a file with the [name] and [content] to the attachments.
      */
+    @Deprecated(
+        "Use ByteReadChannel instead of InputStream",
+        ReplaceWith(
+            "addFile(name, content.toByteReadChannel())",
+            "io.ktor.util.cio.toByteReadChannel"
```
[the remark this answers is not in the document]
Comment (Tmpod) [reply rest/src/main/kotlin/builder/message/create/MessageCreateBuilder.kt:62]: Their documentation is wrong then: https://api.ktor.io/ktor-io/io.ktor.utils.io.jvm.javaio/to-byte-read-channel.html  
I suppose they copied from the CIO variant and forgot to change it.

**Comment by Tmpod** [comment]
Cool!  
I'll be out for most of today, so I'll either submit this later in the
evening or tomorrow.

File: rest/src/main/kotlin/builder/message/create/MessageCreateBuilder.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lukellmann) [reply rest/src/main/kotlin/builder/message/create/MessageCreateBuilder.kt:62]: I don't think it's wrong, they just use another default for the `context` parameter.

**Comment by Tmpod** [comment]
So yeah, now it complains that `_component2` is using the deprecated `inputStream` property. The compiler should see that both are deprecated and not raise a warning, no?  
Anyways, what do you suggest I do? Just have `_component2` mimic `inputStream` or use a `@Suppress`? I assume the latter?

**Comment by Lukellmann** [comment]
> So yeah, now it complains that `_component2` is using the deprecated `inputStream` property. The compiler should see that both are deprecated and not raise a warning, no?  
> Anyways, what do you suggest I do? Just have `_component2` mimic `inputStream` or use a `@Suppress`? I assume the latter?

just slap `@Suppress("DEPRECATION")` on it :)

**Comment by Lukellmann** [comment]
> Also, would like suggestions regarding that test.

I think we actually need to mock a request so that the channel gets consumed. Take a look at `StackTraceRecoveryTest` for an example.

**Comment by Tmpod** [comment]
Right... :facepalm:   
That was extra silly lol

**Comment by Tmpod** [comment]
> > Also, would like suggestions regarding that test.
> 
> I think we actually need to mock a request so that the channel gets consumed. Take a look at `StackTraceRecoveryTest` for an example.

The channel never seems to get consumed in the mock request, which is odd. I'll push what I have, maybe I'm overlooking something obvious again :upside_down_face:

**Comment by Lukellmann** [comment]
> The channel never seems to get consumed in the mock request, which is odd. I'll push what I have, maybe I'm overlooking something obvious again 🙃

Let me check this out and take a look at it.

**Comment by Tmpod** [comment]
Sure thing! Thanks a lot :)

**Comment by Lukellmann** [comment]
> Sure thing! Thanks a lot :)

I PRed the fix here: https://github.com/Tmpod/kord/pull/1

**Comment by Lukellmann** [comment]
I found another problem with using `ByteReadChannel` directly without a `ChannelProvider`:

[Here](https://github.com/kordlib/kord/blob/c1dd6d6c9b6e5596f6f24322da06dddc63a61e58/rest/src/main/kotlin/request/KtorRequestHandler.kt#L56) we retry a request when we got a rate limit. This means we call `handle` again and will rebuild the request [here](https://github.com/kordlib/kord/blob/c1dd6d6c9b6e5596f6f24322da06dddc63a61e58/rest/src/main/kotlin/request/KtorRequestHandler.kt#L44) by creating a [new `MultiPartFormDataContent`](https://github.com/kordlib/kord/blob/c1dd6d6c9b6e5596f6f24322da06dddc63a61e58/rest/src/main/kotlin/request/KtorRequestHandler.kt#L92) from the [already built `request.data`](https://github.com/kordlib/kord/blob/c1dd6d6c9b6e5596f6f24322da06dddc63a61e58/rest/src/main/kotlin/request/Request.kt#L88). Since `ByteReadChannel` can only be read once, the second request will also fail because of this.

This is actually also a problem with `InputStream` that we had before, but now that we refactor this, what should we do?
- add the convenience functions that take `ByteReadChannel` (like it is rn)
- add the convenience functions that take `ByteReadChannel` but add a warning in the docs
- don't have these convenience functions (I would actually prefer this)

Of course rate limit on the exact request that uploads a file might not happen very often, but when it does, I'm pretty sure it'll fail.

**Review by Lukellmann — COMMENTED** [review]
Suggestions for updates to deprecation messages and replacements based on this comment: https://github.com/kordlib/kord/pull/682#issuecomment-1242581969

**Comment by Tmpod** [comment]
Mmm, well spotted.

> This is actually also a problem with `InputStream` that we had before, but now that we refactor this, what should we do?
> - add the convenience functions that take `ByteReadChannel` (like it is rn)
> - add the convenience functions that take `ByteReadChannel` but add a warning in the docs
> - don't have these convenience functions (I would actually prefer this)

I agree with the last one yeah. We can still have `Path` overloads that make providers like  
`ChannelProvider { path.readChannel() }`  
but prevent people from making the mistake of providing a single channel that can potentially cause issues.

File: rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt [hunk 8]
```diff
@@ -32,20 +33,39 @@ public sealed interface MessageModifyBuilder {
...
-        return namedFile
-    }
-
-    public suspend fun addFile(path: Path): NamedFile = withContext(Dispatchers.IO) {
-        addFile(path.fileName.toString(), Files.newInputStream(path))
+    public fun addFile(file: NamedFile) {
+        files = (files ?: mutableListOf()).also {
+            it.add(file)
+        }
     }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt:68]: Why did you add this method? It wasn't there before 🤔

**Review by Lukellmann — COMMENTED** [review]
Looks good overall, just one last question left

File: rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt (hunk 8 above)
Comment (Tmpod) [reply rest/src/main/kotlin/builder/message/modify/MessageModifyBuilder.kt:68]: Was to remove code duplication in the previous implementation, but now it's obsolete cus all other overloads use ChannelProvider. Will remove

**Review by Lukellmann — APPROVED** [review]
I've committed my final touches [here](https://github.com/kordlib/kord/pull/682/commits/872275196b8cea8fd2cf02757225fc8f1ec4026c).

Could you update title and description of the PR (we no longer use `ByteReadChannel` and the test wasn't removed)? Then we would be good to go. Thanks for your efforts :)

**Comment by Tmpod** [comment]
> I've committed my final touches [here](https://github.com/kordlib/kord/pull/682/commits/872275196b8cea8fd2cf02757225fc8f1ec4026c).

Perfect!

> Could you update title and description of the PR

Yep, good point hehe

Thank you for your excellent reviewing! Helped me get back into Kotlin again :3
